File: tools/quality/project_report.py
#!/usr/bin/env python3
"""Frozen non-authoritative Python reference; see docs/quality/python-retention.md.

Generic decisions belong to the released Rust core. No release-approval path.

Check opt-in project configuration and report generic shadow gates."""
from __future__ import annotations

import argparse
import json
from pathlib import Path
import sys

import harness_evidence as evidence
import policy_engine as engine
import project_model as model


def load_config(path, root):
    config = evidence.load_json(path)
    evidence._shape(config, 'project-config.schema.json')
    root = Path(root).resolve()

    def referenced(name):
        relative = model.canonical_path(config[name])
        resolved = (root / relative).resolve()
        evidence.require(resolved.is_relative_to(root), 'config reference escapes project root')
        return evidence.load_json(resolved)

    project, policy = referenced('project_model'), referenced('policy')
    engine.validate_policy(policy, project)
    return project, policy


def report(policy_result, project, policy):
    """Indexes point to one lossless gate table, including raw evidence links."""
    gates = {}
    indexes = {name: {} for name in ('component', 'subject', 'policy', 'status', 'relationship')}
    for position, result in enumerate(policy_result['results']):
        gate_id = f'gate-{position:06d}'
        gates[gate_id] = result
        record = result['record']
        link = record.get('relationship')
        components = ([link['producer'], link['consumer']] if link else
                      [record['component']] if record['component'] else [])
        values = {'component': components, 'subject': [result['subject']],
                  'policy': [result['policy']], 'status': [result['state']],
                  'relationship': [link['id']] if link else []}
        for dimension, keys in values.items():
            for key in keys:
                if key is not None:
                    indexes[dimension].setdefault(key, []).append(gate_id)

    def aggregate(ids):
        results = [engine.GateResult(**gates[i]) for i in ids]
        policies = {r.policy for r in results}
        if not policies:
            return {'state': 'not_applicable', 'blockers': []}
        return engine.aggregate({**policy, 'rules': [r for r in policy['rules']
                                                    if r['id'] in policies]}, results)

    components = {}
    for component in project['components']:
        ids = indexes['component'].get(component['id'], [])
        local = [i for i in ids if 'relationship' not in gates[i]['record']]
        cross = [i for i in ids if 'relationship' in gates[i]['record']]
        components[component['id']] = {'gates': ids, 'aggregate': aggregate(ids),
                                      'local': aggregate(local), 'cross_component': aggregate(cross)}
    return {'schema': 'harness-project-report/v1', 'mode': 'shadow',
            'project': project['id'], 'aggregate': policy_result['aggregate'],
            'components': components, 'gates': gates, 'indexes': indexes,
            'policy_result': policy_result}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('command', choices=('check', 'evaluate'))
    parser.add_argument('--config', type=Path, default=Path('.harness-gate/project.json'))
    parser.add_argument('--root', type=Path, default=Path('.'),
                        help='Root for explicit model/policy references; defaults to working directory')
    parser.add_argument('--output', required=True, type=Path)
    for name in ('evidence', 'expected', 'source-root', 'artifact-root', 'selection'):
        parser.add_argument('--' + name, type=Path)
    parser.add_argument('--reference-only', action='store_true', required=True,
                        help='Acknowledge frozen reference output cannot approve a release')
    args = parser.parse_args()
    try:
        project, policy = load_config(args.config, args.root)
        if args.command == 'check':
            result = {'schema': 'harness-project-config-check/v1', 'mode': 'shadow',
                      'project': project['id'], 'aggregate': {'state': 'pass'}}
        else:
            evidence.require(all((args.evidence, args.expected, args.source_root, args.artifact_root)),
                             'evaluate requires evidence, expected, source-root and artifact-root')
            result = report(engine.evaluate(
                policy, evidence.load_json(args.evidence), project=project,
                expected=evidence.load_json(args.expected), source_root=args.source_root,
                artifact_root=args.artifact_root,
                selection=evidence.load_json(args.selection) if args.selection else None), project, policy)
    except (evidence.MeasurementError, model.ModelError, OSError) as error:
        result = {'schema': 'harness-project-report/v1', 'mode': 'shadow',
                  'aggregate': {'state': 'measurement_error'}, 'error': str(error)}
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(result, indent=2, sort_keys=True) + '\n')
    return 0 if result['aggregate']['state'] == 'pass' else 1


if __name__ == '__main__':
    sys.exit(main())

File: tools/harness-gate/quality-core/project_report.rs
//! Lossless project gate tables and indexes over evaluated generic policy results.
use super::{array, error, policy, string, Result};
use serde_json::{json, Map, Value};
use std::collections::{BTreeMap, BTreeSet};

/// Build the stable shadow report from a validated project/policy and its result.
/// Gate order and evidence links are preserved; participant indexes may share gates.
pub fn report(policy_result: &Value, project: &Value, policy: &Value) -> Result<Value> {
    let mut gates = Map::new();
    let mut indexes: BTreeMap<&str, BTreeMap<String, Vec<String>>> =
        ["component", "subject", "policy", "status", "relationship"]
            .into_iter()
            .map(|name| (name, BTreeMap::new()))
            .collect();
    for (position, result) in array(&policy_result["results"]).iter().enumerate() {
        let id = format!("gate-{position:06}");
        gates.insert(id.clone(), result.clone());
        let record = &result["record"];
        let link = record
            .get("relationship")
            .filter(|v| v.is_object() && !v.as_object().unwrap().is_empty());
        let components = if let Some(link) = link {
            vec![&link["producer"], &link["consumer"]]
        } else if record["component"].as_str().is_some_and(|s| !s.is_empty()) {
            vec![&record["component"]]
        } else {
            vec![]
        };
        for (dimension, keys) in [
            ("component", components),
            ("subject", vec![&result["subject"]]),
            ("policy", vec![&result["policy"]]),
            ("status", vec![&result["state"]]),
            (
                "relationship",
                link.map_or_else(Vec::new, |l| vec![&l["id"]]),
            ),
        ] {
            for key in keys.into_iter().filter(|k| !k.is_null()) {
                indexes
                    .get_mut(dimension)
                    .unwrap()
                    .entry(string(key).into())
                    .or_default()
                    .push(id.clone());
            }
        }
    }
    let aggregate = |ids: &[String]| -> Result<Value> {
        let results: Vec<policy::GateResult> = ids
            .iter()
            .map(|id| {
                serde_json::from_value(gates[id].clone())
                    .map_err(|e| error(format!("invalid gate result: {e}")))
            })
            .collect::<Result<_>>()?;
        let policies: BTreeSet<_> = results.iter().map(|r| r.policy.as_str()).collect();
        if policies.is_empty() {
            return Ok(json!({"state":"not_applicable", "blockers":[]}));
        }
        let mut subset = policy.clone();
        subset["rules"] = array(&policy["rules"])
            .iter()
            .filter(|r| policies.contains(string(&r["id"])))
            .cloned()
            .collect();
        policy::aggregate(&subset, &results)
    };
    let mut components = Map::new();
    for component in array(&project["components"]) {
        let id = string(&component["id"]);
        let ids = indexes["component"].get(id).cloned().unwrap_or_default();
        let (cross, local): (Vec<_>, Vec<_>) = ids
            .iter()
            .cloned()
            .partition(|i| gates[i]["record"].get("relationship").is_some());
        components.insert(
            id.into(),
            json!({"gates":ids, "aggregate":aggregate(&ids)?,
            "local":aggregate(&local)?, "cross_component":aggregate(&cross)?}),
        );
    }
    Ok(
        json!({"schema":"harness-project-report/v1", "mode":"shadow", "project":project["id"],
        "aggregate":policy_result["aggregate"], "components":components, "gates":gates,
        "indexes":indexes, "policy_result":policy_result}),
    )
}
